File: algorithms/medium/best_team_with_no_conflicts.ex
defmodule Solution do
  @spec best_team_score(scores :: [integer], ages :: [integer]) :: integer
  def best_team_score(scores, ages) do
    teams =
      Enum.zip(ages, scores)
      |> Enum.sort_by(fn {a, s} -> {-a, -s} end)
      |> Enum.with_index()
      |> Enum.reduce(%{}, fn {{_, s}, i}, acc ->
        Map.put(acc, i, s)
      end)

    for i <- 0..(map_size(teams) - 1), reduce: {%{}, 0} do
      {dp, ans} ->
        score = teams[i]
        dp = Map.put(dp, i, score)

        for j <- 0..(i - 1)//1, reduce: dp do
          dp ->
            older_score = teams[j]

            if older_score >= score do
              Map.update!(dp, i, &max(&1, dp[j] + score))
            else
              dp
            end
        end
        |> then(fn dp ->
          {dp, max(ans, dp[i])}
        end)
    end
    |> elem(1)
  end
end

File: algorithms/medium/best_team_with_no_conflicts.cpp
class Solution {
public:
    int bestTeamScore(vector<int>& scores, vector<int>& ages) {
      vector<pair<int, int>> teams;
      for (int i = 0; i < scores.size(); i++) {
        teams.push_back({ages[i], scores[i]});
      }
      sort(teams.begin(), teams.end(), greater<pair<int, int>>());
      vector<int> dp(teams.size());
      int ans = 0;
      for (int i = 0; i < teams.size(); i++) {
        dp[i] = teams[i].second;
        for (int j = 0; j < i; j++) {
          if (teams[j].second >= teams[i].second) {
            dp[i] = max(dp[i], dp[j] + teams[i].second);
          }
        }
        ans = max(ans, dp[i]);
      }
      return ans;
    }
};
